src/logic/otp2/controller.ts

```diff
@@ -536,15 +513,131 @@ async function searchTaxiFrontBack(
     }
 }
 
-export type NonTransitMode = 'foot' | 'bicycle' | 'car' | 'bike_rental'
+// TODO 7/12-21: Clean up function signature and types once OTP1 is gone.
+// As we have removed recursion, initialSearchDate can be replaced with searchDate. searchFilter could
+// be kept separate from Otp2SearchParams.
+export async function searchTransit(
+    { initialSearchDate, searchFilter, ...searchParams }: Otp2SearchParams,
+    extraHeaders: { [key: string]: string },
+): Promise<TransitTripPatterns> {
+    const { searchDate, arriveBy } = searchParams
+
+    const getTripPatternsParams = {
+        ...searchParams,
+        modes: filterModesAndSubModes(searchFilter),
+    }
+
+    // We do two searches in parallel here to speed things up a bit. One is a flexible search where
+    // we explicitly look for trips that may include means of transport that has to be booked in advance,
+    // the second is a regular search.
+    const [
+        flexibleResults,
+        [regularTripPatterns, initialMetadata, routingErrors],
+    ] = await Promise.all([
+        searchFlexible(searchParams),
+        getTripPatterns(getTripPatternsParams),
+    ])
+
+    let queries = [
+        ...flexibleResults.queries,
+        getTripPatternsQuery(getTripPatternsParams),
+    ]
+
+    let metadata = initialMetadata
+
+    const noStopsInRangeErrors = routingErrors.filter(
+        ({ code }) => code === RoutingErrorCode.noStopsInRange,
+    )
+
+    // If we have any noStopsInRange errors, we couldn't find a means of transport from where the
+    // traveler wants to start or end the trip. Try to find an option using taxi for those parts instead.
+    let taxiTripPatterns: Otp2TripPattern[] = []
+    if (noStopsInRangeErrors.length > 0) {
+        const noFromStopInRange = noStopsInRangeErrors.some(
+            (e) => e.inputField === 'from',
+        )
+        const noToStopInRange = noStopsInRangeErrors.some(
+            (e) => e.inputField === 'to',
+        )
+
+        const taxiResults = await searchTaxiFrontBack(searchParams, {
+            access: noFromStopInRange,
+            egress: noToStopInRange,
+        })
+        taxiTripPatterns = taxiResults.tripPatterns
+        queries = [...taxiResults.queries]
+    }
+
+    let tripPatterns = regularTripPatterns
```

Comment:
There is something really icky about having tripPatterns as a mutable. At one point, after 'searchBeforeFlexible' it is even completely overwritten (which is fine because if we arrive there it is actually empty) and at other times it is amended to (though not by mutation). Knowing what is actually returned and in what order is not easy to see.